MachineLearning/data/aggregate/aggregate_in_time.py

```diff
@@ -6,42 +6,116 @@
 """
 import numpy as np
 import pandas as pd
+import datetime as dt
 
 def aggregate(input_file_dust, input_file_weather):
-    
     with open(input_file_weather) as weather:
-        data = pd.read_csv(weather, sep=';')
+        data = pd.read_csv(weather, sep=',').drop_duplicates(subset='dt',keep='first').set_index('dt') # indexing by unix timestamp for faster lookups
     filtered_data = data[['dt_iso', 'pressure', 'humidity', 'wind_speed', 'wind_deg']]
-    filtered_data = filtered_data.head(31*24)
 
     with open(input_file_dust) as dust:
         data = pd.read_csv(dust, sep=';')
     filtered_dust_data = data[['sensor_id', 'timestamp', 'P1', 'P2']]
-    sensor_id = filtered_dust_data[['sensor_id']].head(31*24)
     n_rows = filtered_dust_data.shape[0]
-    p1_av = []
-    p2_av = []
-    n_measurements_per_day = 21
-    print(n_measurements_per_day)
-    p1_data = data.P1
-    p2_data = data.P2
-        
-    n_aggregations = n_rows//n_measurements_per_day
-    print(n_aggregations)
-    offset = 0
-    for i in range(n_aggregations):
-        p1_av.append(np.mean(p1_data[offset:offset+n_measurements_per_day]))
-        p2_av.append(np.mean(p2_data[offset:offset+n_measurements_per_day]))
-        offset = (i+1)*n_measurements_per_day
-            
-    p1_av.append(np.mean(p1_data[offset:]))
-    p2_av.append(np.mean(p2_data[offset:]))
-        
-    filtered_data['node_id'] = sensor_id
-    filtered_data['P1'] = p1_av[0:31*24] 
-    filtered_data['P2'] = p2_av[0:31*24] 
+
+    # the main issue here is that we have multiple sensors in the dust input file
+    # we are using a per-sensor accumulator to compute the running counts and then averages
+    p1_accumulator = {}
+    p2_accumulator = {}
+    length_accumulator = {}
+
+    sensors = filtered_dust_data[['sensor_id']]
+    timestamps = filtered_dust_data[['timestamp']]
+    p1s = filtered_dust_data[['P1']]
+    p2s = filtered_dust_data[['P2']]
+
+    current_time = None # for detection
+
+    p1_avg = []
+    p2_avg = []
+    sensors_ids = []
+    dt_isos = []
+    pressures = []
+    humidities = []
+    wind_speeds = []
+    wind_degs = []
+
+    for i in range(n_rows):
+        sensor = sensors.at[i, 'sensor_id']
+        p1 = p1s.at[i, 'P1']
+        p2 = p2s.at[i, 'P2']
+        time_string = timestamps.at[i, 'timestamp']
+        time = pd.to_datetime(time_string, infer_datetime_format=True, utc=True)
+
+        if (current_time is not None) and (time.hour > current_time.hour):
```

Comment:
If I understand correctly, you are resetting the dictionary at the beginning of each day. In that case should not the condition be inverted: time.hour < current_time.hour. Since during the day, the hour is increasing and when the day switches to another one, the hour resets to 0. The current_time.hour is always the one from the measurement before, right?

Reply:
We are resetting at each change in hour. But you are correct the condition is wrong - it should read `time.hour != current_time.hour`. That will cover all the cases when we need to reset.